balancer_conn_wrappers.go

```diff
@@ -414,3 +418,67 @@ func (acbw *acBalancerWrapper) getAddrConn() *addrConn {
 	defer acbw.mu.Unlock()
 	return acbw.ac
 }
+
+var errSubConnNotReady = status.Error(codes.Unavailable, "SubConn not currently connected")
+
+// NewStream begins a streaming RPC on the addrConn.  If the addrConn is not
+// ready, returns errSubConnNotReady.
+func (acbw *acBalancerWrapper) NewStream(ctx context.Context, desc *StreamDesc, method string, opts ...CallOption) (ClientStream, error) {
+	transport := acbw.ac.getReadyTransport()
+	if transport == nil {
+		return nil, errSubConnNotReady
+	}
+	return newNonRetryClientStream(ctx, desc, method, transport, acbw.ac, opts...)
+}
+
+// Invoke performs a unary RPC.  If the addrConn is not ready, returns
+// errSubConnNotReady.
+func (acbw *acBalancerWrapper) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...CallOption) error {
+	cs, err := acbw.NewStream(ctx, unaryStreamDesc, method, opts...)
+	if err != nil {
+		return err
+	}
+	if err := cs.SendMsg(args); err != nil {
+		return err
+	}
+	return cs.RecvMsg(reply)
+}
+
+// producerData stores a producer, a ref counting mechanism, and a close
+// function to be called when the producer no longer has any references.
+type producerData struct {
+	producer balancer.Producer
+	refs     int
+	close    func()
+}
+
+func (acbw *acBalancerWrapper) GetOrBuildProducer(pb balancer.ProducerBuilder) (balancer.Producer, func()) {
+	acbw.mu.Lock()
+	defer acbw.mu.Unlock()
+
+	// Look up existing producer from this builder.
+	pData := acbw.producers[pb]
+	if pData == nil {
+		// Not found; create a new one and add it to the producers map.
+		p, close := pb.Build(acbw)
+		pData = &producerData{producer: p, close: close}
+		acbw.producers[pb] = pData
+	}
+	// Account for this new reference.
+	pData.refs++
+
+	// Return a cleanup function wrapped in a sync.Once to remove this
+	// reference and delete the producerData from the map if the total
+	// reference count goes to zero.
+	unref := func() {
+		acbw.mu.Lock()
+		pData.refs--
+		if pData.refs == 0 {
```

Comment:
> keep open streams even if we have no SubConns connected to the server a stream is associated with

This doesn't make sense to me.  Streams live on connections to servers, and if the connection goes away, the stream is lost.

Or do you mean, if all the ORCA(/etc) subscribers go away on a SubConn, keep the producer around for a few minutes in case a new one reappears?  We could do that technically, but I don't think it's worth it, and I wouldn't want to do it at the grpc level - the plugin should have full control over such things.  The ORCA producer could delay unsubscription requests instead.

Reply:
The latter, but I agree it's too complicated technically to be worth it.